lib/Conversion/PolynomialToStandard/PolynomialToStandard.cpp

```diff
@@ -265,32 +307,66 @@ struct ConvertMulScalar : public OpConversionPattern<MulScalarOp> {
     ImplicitLocOpBuilder b(op.getLoc(), rewriter);
     auto tensorType = cast<RankedTensorType>(adaptor.getPolynomial().getType());
     Value scalar = adaptor.getScalar();
-    if (scalar.getType().getIntOrFloatBitWidth() <
-        tensorType.getElementTypeBitWidth()) {
-      scalar = b.create<arith::ExtSIOp>(tensorType.getElementType(), scalar)
-                   .getResult();
-    } else if (scalar.getType().getIntOrFloatBitWidth() >
-               tensorType.getElementTypeBitWidth()) {
-      scalar = b.create<arith::TruncIOp>(tensorType.getElementType(), scalar)
-                   .getResult();
-    }
+    // MulScaparOp verifier enforces that the input has the same type as the
+    // polynomial ring's coefficient type.
     auto tensor = b.create<tensor::SplatOp>(tensorType, scalar);
-    rewriter.replaceOpWithNewOp<arith::MulIOp>(op, adaptor.getPolynomial(),
-                                               tensor);
+    auto mulOp = b.create<arith::MulIOp>(adaptor.getPolynomial(), tensor);
+    Operation *finalOp = mulOp;
+
+    RingAttr ring =
+        cast<PolynomialType>(op.getPolynomial().getType()).getRing();
+    if (ring.getCoefficientModulus()) {
+      APInt mod = ring.getCoefficientModulus().getValue();
+      if (needToMod(ring.getCoefficientType(), mod)) {
+        auto modValue = b.create<arith::ConstantOp>(
+            tensorType.getElementType(),
+            IntegerAttr::get(
+                tensorType.getElementType(),
+                mod.zextOrTrunc(
+                    tensorType.getElementType().getIntOrFloatBitWidth())));
+        finalOp = b.create<arith::RemSIOp>(adaptor.getPolynomial(), modValue);
+      }
+    }
+
+    rewriter.replaceOp(op, finalOp);
     return success();
   }
 };
 
+bool hasCyclicModulus(MonicMonomialMulOp op) {
```

Comment:
nit: This is OK naming but could you add a quick docstring about the form of this polynomial since it doesn't quite line up exactly with the name? I think it qualifies as a cyclic polynomial (one that's invariant under symmetric group operations on the coefficients since it only has one degree) but it's also not a cyclotomic polynomial (since x^n - 1 may not be irreducible).

Reply:
Done.